robocop/checkers/misc.py

```diff
@@ -258,3 +258,52 @@ def visit_LibraryImport(self, node):  # noqa
         if not node.name:
             return
         self.libraries.append(node)
+
+
+class UnusedVariableChecker(VisitorChecker):
+    """ Checker for unused variable. """
+    rules = {
+        "0912": (
+            "variable-not-used",
+            "The variable is assigned but not used in this scope",
+            RuleSeverity.WARNING
+        )
+    }
+
+    def __init__(self):
+        self.variable_list = []
+        super().__init__()
+
+    def visit_KeywordCall(self, node):  # noqa
+        for child in node.data_tokens:
+            if child.type == Token.ASSIGN:
+                self.variable_list.append([child, False])
+            elif child.type == Token.ARGUMENT:
+                self.use_variable(child)
+
+    def visit_Keyword(self, node):  # noqa
+        self.variable_list.clear()
+        self.generic_visit(node)
+        self.check_unused()
+
+    def visit_TestCase(self, node):  # noqa
+        self.variable_list.clear()
+        self.generic_visit(node)
+        self.check_unused()
+
+    def use_variable(self, node):
+        node_nomalized = self.normalize(node.value)
+        for i, token in enumerate(self.variable_list):
```

Comment:
Instead of storing multiple copies of the same variable (if it's used twice) and renornalizing whole list of variables, maybe it would be better to save it to set or dict (and already normalized)?

Reply:
If I change `list` to `dict`, the rule would only catch once in the following situation. Is this expect result correct?
```robotframework
** Settings ***
Documentation    This is suite doc


*** Test Cases ***
Test
    [Documentation]   doc
    ${var}    Keyword
    ${var}    Keyword 2
```

Reply:
If I change `list` to `dict`, the rule would only catch once in the following situation. Is this expect result correct?
```robotframework
** Settings ***
Documentation    This is suite doc


*** Test Cases ***
Test
    [Documentation]   doc
    ${var}    Keyword
    ${var}    Keyword 2
```

Reply:
In this case it should work as dict - because you're redefining variable. 

ie:
```
    ${var}    Keyword
    ${var}    Keyword 2
    Log  ${var}
```
Your current checker will not detect that first ${var} was never used and just redifined. 

and in this case
```
    ${var}    Keyword
    ${var}    Keyword 2
```
If you're redefining not used variable - it should throw warning. Then it should throw warning on second ${var} because it's also not used anywhere.